File: leetcode2025/2025_12/leetcode2025-12-23.py
from typing import List
from bisect import *

class Solution:
    def maxTwoEvents(self, events: List[List[int]]) -> int:
        # 贪心，对于每一个活动，如果它是作为第二个活动被选择，那么上一个活动的结束时间一定是小于当前活动的开始时间的，因此可以将所有活动按照结束时间顺序排列，二分查找求出最后一个可能的位置，用前缀最大值求解即可
        n = len(events)
        events.sort(key = lambda x:x[1])
        endTime = [x[1] for x in events]
        preMax = [0]*(n+1)
        for i in range(n):
            preMax[i+1] = max(preMax[i], events[i][2])
        
        res = preMax[-1]
        for i in range(n):
            end = events[i][0]-1
            indx = bisect_left(endTime, end+1)-1
            res = max(res, events[i][2]+preMax[indx+1])
        return res

if __name__ == "__main__":
    s = Solution()
    events = [[1,3,2],[4,5,2],[2,4,3]]
    print(s.maxTwoEvents(events))

File: leetcode2025/2025_12/leetcode2025-12-23.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxTwoEvents(vector<vector<int>>& events) {
        sort(events.begin(), events.end(), [](auto& a, auto& b){
            return (a[1] < b[1]);
        });
        int n = events.size();
        vector<int> endTime(n);
        vector<int> preMax(n+1, 0);
        for (int i=0; i<n; i++){
            endTime[i] = events[i][1];
            preMax[i+1] = max(preMax[i], events[i][2]);
        }
        
        int res = preMax[n-1];
        for (int i=0; i<n; i++){
            int end = events[i][0]-1;
            int indx = lower_bound(endTime.begin(), endTime.end(), end+1)-endTime.begin()-1;
            res = max(res, events[i][2]+preMax[indx+1]);
        }
        return res;
    }
};
// 这是协同测试的小例子
// This is an example of cooperate test.
// hahah!
// 你要干嘛
// 我只是觉得好玩而已
// 哦
// 你赶快写完吧
// 现在就好了哦
// 呵呵

int main(){
    Solution s;
    vector<vector<int>> events = {{1,3,2},{4,5,2},{2,4,3}};
    cout << s.maxTwoEvents(events) << endl;
}
